files, = glob_wildcards("data/{sample}.fastq.gz")


include: "rules/ncrna_workflow.smk"
include: "rules/statistics.smk"
include: "rules/tRF_workflow.smk"


route=config.get("route","bowtie2")



if route == "bowtie2":
	rule all:
		input:
			#expand("results/count_tables/{gene}.tsv",gene=gene),
			expand(["results/count_tables/" + x + ".tsv" for x in ["lncRNA","misc_RNA","protein_coding","snoRNA","snRNA","scaRNA"]]),
			"results/file_statistics.csv",
			expand("analyses/fastqc/{sample}.trimmed.FastQC/{sample}.trimmed_fastqc.zip",sample=files),
			"results/multiqc_report.html",
			"results/count_tables/tRF.tsv",
			"results/mirtrace/mirtrace-report.html"
			
elif route == "bowtie1":
	print("Bowtie1 module has been selected.")
	rule all_bowtie1:
		input:
			expand("analyses/bowtie1_mappings/{sample}.sorted.bam",sample=files)


elif route == "stats" or route == "statistics":
	print("Statistics module has been selected.")
	rule statistics:
		input:
			expand("results/statistics/{gene}.tsv",gene=["gencode.uniq.exon","piRNA.uniq","tRNA.uniq"])
	

else:
	print("Please select a correct route...")
	print("bowtie1/bowtie2/stats")

